File: java/logbrew-java/src/main/java/co/logbrew/sdk/LogBrewOpenTelemetry.java
package co.logbrew.sdk;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanContext;
import io.opentelemetry.context.Context;
import java.util.Optional;

/**
 * Optional OpenTelemetry API helpers for apps that already install OpenTelemetry.
 *
 * <p>The helpers copy only valid trace ID, span ID, and trace flags into a
 * LogBrew child trace context. They do not own OpenTelemetry SDK providers,
 * exporters, processors, attributes, links, events, baggage, or tracestate.</p>
 */
public final class LogBrewOpenTelemetry {
    private LogBrewOpenTelemetry() {
    }

    /**
     * Copies the current OpenTelemetry span into a LogBrew child context.
     */
    public static Optional<LogBrewTraceContext> traceContextFromCurrentSpan() {
        return traceContextFromSpan(Span.current());
    }

    /**
     * Copies the current OpenTelemetry span into a LogBrew child context with an app-owned span ID.
     */
    public static Optional<LogBrewTraceContext> traceContextFromCurrentSpan(String spanId) {
        return traceContextFromSpan(Span.current(), spanId);
    }

    /**
     * Copies a span from an explicit OpenTelemetry context into a LogBrew child context.
     */
    public static Optional<LogBrewTraceContext> traceContextFromContext(Context context) {
        if (context == null) {
            return Optional.empty();
        }
        return traceContextFromSpan(Span.fromContext(context));
    }

    /**
     * Copies a span from an explicit OpenTelemetry context into a LogBrew child context.
     */
    public static Optional<LogBrewTraceContext> traceContextFromContext(Context context, String spanId) {
        if (context == null) {
            return Optional.empty();
        }
        return traceContextFromSpan(Span.fromContext(context), spanId);
    }

    /**
     * Copies an OpenTelemetry span into a LogBrew child context.
     */
    public static Optional<LogBrewTraceContext> traceContextFromSpan(Span span) {
        if (span == null) {
            return Optional.empty();
        }
        return traceContextFromSpanContext(span.getSpanContext());
    }

    /**
     * Copies an OpenTelemetry span into a LogBrew child context with an app-owned span ID.
     */
    public static Optional<LogBrewTraceContext> traceContextFromSpan(Span span, String spanId) {
        if (span == null) {
            return Optional.empty();
        }
        return traceContextFromSpanContext(span.getSpanContext(), spanId);
    }

    /**
     * Copies an OpenTelemetry span context into a LogBrew child context.
     */
    public static Optional<LogBrewTraceContext> traceContextFromSpanContext(SpanContext spanContext) {
        if (spanContext == null || !spanContext.isValid()) {
            return Optional.empty();
        }
        return Optional.of(LogBrewTraceContext.fromTraceparent(parentTraceparent(spanContext)));
    }

    /**
     * Copies an OpenTelemetry span context into a LogBrew child context with an app-owned span ID.
     */
    public static Optional<LogBrewTraceContext> traceContextFromSpanContext(
        SpanContext spanContext,
        String spanId
    ) {
        if (spanContext == null || !spanContext.isValid()) {
            return Optional.empty();
        }
        return Optional.of(LogBrewTraceContext.fromTraceparent(parentTraceparent(spanContext), spanId));
    }

    private static String parentTraceparent(SpanContext spanContext) {
        return Traceparent.create(
            spanContext.getTraceId(),
            spanContext.getSpanId(),
            spanContext.getTraceFlags().asHex()
        );
    }
}

File: kotlin/logbrew-kotlin/src/main/kotlin/co/logbrew/sdk/LogBrewOpenTelemetry.kt
package co.logbrew.sdk

private const val OTEL_SPAN_CLASS = "io.opentelemetry.api.trace.Span"
private const val OTEL_SPAN_CONTEXT_CLASS = "io.opentelemetry.api.trace.SpanContext"
private const val OTEL_TRACE_FLAGS_CLASS = "io.opentelemetry.api.trace.TraceFlags"
private const val OTEL_CONTEXT_CLASS = "io.opentelemetry.context.Context"

object LogBrewOpenTelemetry {
    @JvmStatic
    fun spanContextFromCurrentSpan(): LogBrewOpenTelemetrySpanContext? {
        val span = invokeStaticNoArg(OTEL_SPAN_CLASS, "current") ?: return null
        return spanContextFromSpan(span)
    }

    @JvmStatic
    fun traceContextFromCurrentSpan(): LogBrewTraceContext? = spanContextFromCurrentSpan()?.let(LogBrewTrace::fromOpenTelemetrySpanContext)

    @JvmStatic
    fun spanContextFromSpan(span: Any?): LogBrewOpenTelemetrySpanContext? {
        val otelSpan = instanceOf(OTEL_SPAN_CLASS, span) ?: return null
        val spanContext = invokeNoArg(otelSpan, OTEL_SPAN_CLASS, "getSpanContext") ?: return null
        return spanContextFromRawSpanContext(spanContext)
    }

    @JvmStatic
    fun traceContextFromSpan(span: Any?): LogBrewTraceContext? = spanContextFromSpan(span)?.let(LogBrewTrace::fromOpenTelemetrySpanContext)

    @JvmStatic
    fun spanContextFromContext(context: Any?): LogBrewOpenTelemetrySpanContext? {
        val span = spanFromContext(context) ?: return null
        return spanContextFromSpan(span)
    }

    @JvmStatic
    fun traceContextFromContext(context: Any?): LogBrewTraceContext? =
        spanContextFromContext(context)?.let(LogBrewTrace::fromOpenTelemetrySpanContext)

    private fun spanContextFromRawSpanContext(spanContext: Any): LogBrewOpenTelemetrySpanContext? {
        val otelSpanContext = instanceOf(OTEL_SPAN_CONTEXT_CLASS, spanContext) ?: return null
        val valid = invokeNoArg(otelSpanContext, OTEL_SPAN_CONTEXT_CLASS, "isValid") as? Boolean ?: return null
        if (!valid) {
            return null
        }
        val traceId = invokeNoArg(otelSpanContext, OTEL_SPAN_CONTEXT_CLASS, "getTraceId") as? String ?: return null
        val spanId = invokeNoArg(otelSpanContext, OTEL_SPAN_CONTEXT_CLASS, "getSpanId") as? String ?: return null
        val traceFlags = traceFlagsHex(invokeNoArg(otelSpanContext, OTEL_SPAN_CONTEXT_CLASS, "getTraceFlags")) ?: return null
        return LogBrewOpenTelemetrySpanContext.create(traceId, spanId, traceFlags)
    }

    private fun spanFromContext(context: Any?): Any? {
        if (context == null) {
            return null
        }
        val contextClass =
            runCatching {
                Class.forName(OTEL_CONTEXT_CLASS)
            }.getOrNull() ?: return null
        if (!contextClass.isInstance(context)) {
            return null
        }
        return runCatching {
            Class
                .forName(OTEL_SPAN_CLASS)
                .getMethod("fromContext", contextClass)
                .invoke(null, context)
        }.getOrNull()
    }

    private fun traceFlagsHex(traceFlags: Any?): String? {
        val otelTraceFlags = instanceOf(OTEL_TRACE_FLAGS_CLASS, traceFlags) ?: return null
        val explicit = invokeNoArg(otelTraceFlags, OTEL_TRACE_FLAGS_CLASS, "asHex") as? String
        if (explicit != null) {
            return explicit
        }
        val sampled = invokeNoArg(otelTraceFlags, OTEL_TRACE_FLAGS_CLASS, "isSampled") as? Boolean ?: return null
        return if (sampled) "01" else "00"
    }

    private fun instanceOf(
        className: String,
        value: Any?,
    ): Any? {
        if (value == null) {
            return null
        }
        return runCatching {
            val targetClass = Class.forName(className)
            if (targetClass.isInstance(value)) value else null
        }.getOrNull()
    }

    private fun invokeStaticNoArg(
        className: String,
        methodName: String,
    ): Any? =
        runCatching {
            Class.forName(className).getMethod(methodName).invoke(null)
        }.getOrNull()

    private fun invokeNoArg(
        target: Any,
        className: String,
        methodName: String,
    ): Any? =
        runCatching {
            Class.forName(className).getMethod(methodName).invoke(target)
        }.getOrNull()
}
